x-pack/plugins/endpoint/server/routes/whitelist.ts

```diff
@@ -0,0 +1,346 @@
+ */
+export function registerWhitelistRoutes(router: IRouter, endpointAppContext: EndpointAppContext) {
+  router.get(
+    {
+      path: allowlistBaseRoute,
+      validate: {},
+      options: { authRequired: true },
+    },
+    handleWhitelistGet
+  );
+
+  router.get(
+    {
+      path: '/api/endpoint/manifest',
+      validate: {},
+      options: { authRequired: true },
+    },
+    handleWhitelistManifest
+  );
+  router.get(
+    {
+      path: `${allowlistBaseRoute}/download/{hash}`,
+      validate: {
+        params: schema.object({
+          hash: schema.string(),
+        }),
+      },
+      options: { authRequired: true },
+    },
+    handleWhitelistDownload
+  );
+
+  router.post(
+    {
+      path: allowlistBaseRoute,
+      validate: {
+        body: schema.object({
+          comment: schema.maybe(schema.string()), // Optional comment explaining reason for whitelist
+          event_types: schema.arrayOf(schema.string()),
+          file_path: schema.maybe(schema.string()),
+          signer: schema.maybe(schema.string()),
+          sha256: schema.maybe(schema.string()),
+          dismiss: schema.boolean({ defaultValue: false }), // Boolean determining if we dismiss all alerts that match this
+        }),
+      },
+      options: { authRequired: true },
+    },
+    handleWhitelistPost
+  );
+
+  router.delete(
+    {
+      path: allowlistBaseRoute,
+      validate: {
+        body: schema.object({
+          whitelist_id: schema.string(),
+        }),
+      },
+      options: { authRequired: true },
+    },
+    handleWhitelistItemDeletion
+  );
+}
+
+/**
+ * Handles the POST request for whitelist additions
+ */
+async function handleWhitelistPost(context: RequestHandlerContext, req, res) {
+  try {
+    const eventTypes: string[] = req.body.eventTypes;
+    const whitelistAttributeMap: Record<string, string> = {
+      file_path: 'malware.file.path',
+      acting_process_path: 'actingProcess.file.path',
+      signer: 'malware.file.signer',
+      sha256: 'malware.file.hashes.sha256',
+    };
+    const newRules: WhitelistRule[] = [];
+    Object.keys(whitelistAttributeMap).forEach(k => {
+      if (req.body[k]) {
+        newRules.push({
+          comment: req.body.comment || '',
+          event_types: eventTypes,
+          entry_type: 'simple',
+          entry: {
+            comparison: 'equality',
+            value: req.body[k],
+            apply_to: whitelistAttributeMap[k],
+          },
+        });
+      }
+    });
+
+    // Don't index an empty list if no rules could be created from the request
+    if (newRules.length === 0) {
+      return res.badRequest({
+        error: i18n.translate('allowlist.badrequest', {
+          defaultMessage: 'No allowlist rules could be created from request.',
+        }),
+      });
+    }
+
+    // Check that the global whitelist size limit won't be reached
+    if (newRules.length + whitelistCount > whitelistLimit) {
```

Comment:
This is a particularly racy check since `whitelistCount` is based on the last cache update.  We could instead use sequential document IDs when storing the whitelist rules with the `_create` API. Then when inserting a rule first get the live list of rules from elasticsearch, grab the maximum doc ID, and attempt to insert the new rule with the next ID as long as it won't exceed the size limit.  If it fails because a doc with that ID was inserted by a concurrent update we can retry, starting by getting the list of rules again.

Reply:
That would work if the whitelist rules were not able to be deleted. If we created 100 rules, and then a user deletes 50, the ids of the remaining will not change, right?

Reply:
The ids won't change, but we'd still be basing the ID for new whitelist entries on the document with the highest ID so it would maintain the ordering.

Theres still a possible race condition since the operation isn't atomic so you could conceivably create/delete any number of rules in between the check and the insert, but I think the sequential IDs would make it very unlikely to exceed the limit.  Is there another issue with sequential IDs caused by deleting rules?

Reply:
Spoke offline about this. We're going to continue with the current approach but instead of getting the `whitelistCount` from the cache, we will get the current count from ES before insert.